### ISSUE TITLE
Should AppSink callback be lifetime `'static`?

### ISSUE THREAD

**Issue by emdash** [sdroege/gstreamer-rs#127]
https://github.com/sdroege/gstreamer-rs/blob/06bd23f8338724b9baf99d82a267e561e2f16623/gstreamer-app/src/auto/app_sink.rs#L186

By declaring this static, any closure passed to the callback can only capture references with static lifetimes. How do I get data out of the callback?

**Comment by sdroege** [comment]
What's your use-case exactly?

Generally, signals can be emitted from arbitrary places and not just from the scope where you connect to it so all data passed to it must not have references to any stack frames. You would usually `Box` your data in one way or another, or use a `Arc<Mutex<T>>` or similar

**Comment by emdash** [comment]
First thing: I'm porting a prototype from Python to Rust.I just want to store the last frame of video received so it can be painted to the screen on the next update. The use-case requires that some

**Comment by emdash** [comment]
TL; DR: we can't just use an `ximagesink`. When we get a `"draw"` event, I need to have the last frame of video available as a `cairo::Surface` so it can be drawn immediately -- with arbitrary transform and clipping applied, and with custom overlays and text on top. A blocking call to `pull_sample()` on the UI thread to get the current frame seems like a bad idea.

Background: I am working on an embedded application in which the user needs to scan QR codes. This is an embedded full-screen UI. Everything is contained within a single `DrawingArea`, and all the graphics are custom. Video from the camera is shown to provide feedback (it's too hard for them otherwise). I also need to be able to cater to the whims of the CTO and UX designer ;). Software rendering with Cairo might seem crazy, but a Python MVP demonstrated an acceptable frame-rate, and we have some unrleated issues with the graphics drivers that take GL off the table.

**Comment by sdroege** [comment]
Short version, I'm on my phone currently.

Look at the appsink example. You want the new-sample callback and from there dispatch to the main thread. Either the Sample or convert to a Cairo image surface there (see my Cairo PR from today, you can directly create one from a write MappedBuffer).

**Comment by emdash** [comment]
I have looked at the AppSink example, but all it does is print to the console... What I want to do is store data in field of a QRCodeScanner struct, because I don't want this QR scanning stuff spread all over everywhere...

**Comment by sdroege** [comment]
Then you need to store that struct on the heap in one way or another (probably `Arc<Mutex<QRCodeScanner>>` and then store the `gst::Sample` or directly a `cairo::ImageSurface` in there.

**Comment by sdroege** [comment]
(closing as this is unrelated to anything that can be improved in the bindings and more a general Rust question, but feel free to ask for more details here if something is still missing)

**Comment by fengalin** [comment]
@emdash,  you can take a look at [this](https://github.com/fengalin/media-toc/blob/52cf2726e7bff9e0298b8ddc3f836d6204b6f9e3/src/media/playback_context.rs#L477) for an example of the solution suggested by @sdroege.

Note: due to unrelated constraints, this doesn't match the implementation used in the latest master of this project, so stay at this commit level if you want to navigate in the code.

**Comment by sdroege** [comment]
Also https://github.com/gtk-rs/cairo/pull/204 is the Cairo PR I meant above. You can basically do this now

```rust
        // make sure the buffer is writable
        let buffer = buffer.into_mapped_buffer_writable().unwrap();
        // Get format/width/height/stride from a gst_video::VideoInfo. stride is video_info.stride()[0]
        // cairo::Format::Argb32 is gst_video::VideoFormat::Bgra
        // Rgb24 is Bgrx
        let surface = cairo::ImageSurface::create_for_data(buffer, format, width, height, stride);
```

**Comment by emdash** [comment]
No I think that explains a lot. Those examples are really helpful, and honestly something that's hard to find. Most Rust example code I've seen is toy code that does not demonstrate how to structure a medium-sized application properly. I also didn't know about the `lazy-static` crate.
